### PR TITLE
plugin/file: Fix in wrong answers returned when wildcard and concrete records exist

### PR DESCRIPTION
**Pull request by xdu31**
### 1. Why is this pull request needed and what does it do?
It fixes in file plugin wrong answers returned when wildcard and concrete records exist:
<pre>
; example.org test file with wildcard corner cases
$TTL 3600
example.org.		IN	SOA	sns.dns.icann.org. noc.dns.icann.org. 2015082541 7200 3600 1209600 3600  
example.org.		IN	NS	b.iana-servers.net.  
*.d.example.org.        IN      A       127.0.1.53  
r.c.d.example.org.      IN      A       127.0.1.52
</pre>
As per RFC 4592, the query `something.r.c.d.example.com. , A` should not match the wildcard record `*.d.example.com. , A` as `r.c.d.example.com. , A` exists. It is considered the closest encloser, and since there is neither something domain nor a wildcard record under it, the answer has to be NXDOMAIN.
CoreDNS exhibit the same behavior even for queries like `something.c.d.example.com. , A`. The expected response is NXDOMAIN, but CoreDNS returns the wildcard record.

### 2. Which issues (if any) are related?
https://github.com/coredns/coredns/issues/4290

### 4. Does this introduce a backward incompatible change or deprecation?
No

### CODE REVIEW

**Comment by chrisohaver** [comment]
Thanks! In the test case zone `cornerCasesWildcard `, `something.c.d.example.org.` should also be NXDOMAIN, since `c.d.example.org.` is a non-wild empty non-terminal. Can you add this test case? Its a different case than the `something.r.c.d.example.org.` test case, since  `r.c.d.example.org.` is not empty.

File: plugin/file/lookup.go
Comment (miekg) [inline plugin/file/lookup.go]: this comment is a bit terse to explain the previous if statement, please expand a bit

File: plugin/file/lookup.go
Comment (miekg) [inline plugin/file/lookup.go]: superfluous change

File: plugin/file/lookup.go
Comment (miekg) [inline plugin/file/lookup.go]: is `found` actually still used?

**Comment by miekg** [comment]
can you more carefully explain (in the commit mesg), why foundCe/found and ce our now necessary ? Current msg is rather sparse or github is playing tricks on me

File: plugin/file/lookup.go
Comment (xdu31) [reply plugin/file/lookup.go]: > is `found` actually still used?

Yes, in a few other places

https://github.com/coredns/coredns/blob/d6d8a9974bc45855a7d8a89e9e452cc9b6301d0e/plugin/file/lookup.go#L95

https://github.com/coredns/coredns/blob/d6d8a9974bc45855a7d8a89e9e452cc9b6301d0e/plugin/file/lookup.go#L162

https://github.com/coredns/coredns/blob/d6d8a9974bc45855a7d8a89e9e452cc9b6301d0e/plugin/file/lookup.go#L167

https://github.com/coredns/coredns/blob/d6d8a9974bc45855a7d8a89e9e452cc9b6301d0e/plugin/file/lookup.go#L250

File: plugin/file/lookup.go
Comment (xdu31) [reply plugin/file/lookup.go]: Hi @miekg This is the case described in https://github.com/coredns/coredns/issues/4290, when the record "something.r.c.d.example.com" is matching "*.d.example.com" wildcard, "r.c.d.example.com", its closest encloser (CE) can be found, however, its CE is not directly under the wildcard (cntLabel(CE) + 1 > cntLabel(wildcard)), the record needs to be resolved as nxdomain.

**Comment by xdu31** [comment]
> Thanks! In the test case zone `cornerCasesWildcard `, `something.c.d.example.org.` should also be NXDOMAIN, since `c.d.example.org.` is a non-wild empty non-terminal. Can you add this test case? Its a different case than the `something.r.c.d.example.org.` test case, since `r.c.d.example.org.` is not empty.

Hi @chrisohaver @SivaKesava1 Should`something.c.d.example.org.` also return NXDOMAIN? It looks to be the same case as `else.something.d.example.org`, it doesn't have a encloser under the wildcard `*.d.example.org` and is a multi-segment wildcard completion.

**Comment by SivaKesava1** [comment]
> > Thanks! In the test case zone `cornerCasesWildcard `, `something.c.d.example.org.` should also be NXDOMAIN, since `c.d.example.org.` is a non-wild empty non-terminal. Can you add this test case? It's a different case than the `something.r.c.d.example.org.` test case, since `r.c.d.example.org.` is not empty.
> 
> Hi @chrisohaver @SivaKesava1 Should`something.c.d.example.org.` also return NXDOMAIN? It looks to be the same case as `else.something.d.example.org`, it doesn't have a encloser under the wildcard `*.d.example.org` and is a multi-segment wildcard completion.
 
Yes, it should be an NXDOMAIN. Consider the domain names of the records as a tree, then follow the tree label by label from the root and stop when you can go further due to a non-match. If the node at which you stopped has a wildcard child, then you will perform wildcard synthesis; otherwise, you would return an NXDOMAIN.

```
    org
      |
   example
      |
      d
     / \
    c   *
    |
    r
```

When doing the matching, `something.c.d.example.org.` stops at the `c` node, and it does not have a wildcard child, so an NXDOMAIN. For `else.something.d.example.org,` it stops the matching at the `d` node, and it has a wildcard child, so it is a wildcard synthesis.
I am not aware of the internal data structures used by CoreDNS, so I can not elaborate further, but the above explanation is based on how RFCs view the zone file.

**Comment by xdu31** [comment]
> > > Thanks! In the test case zone `cornerCasesWildcard `, `something.c.d.example.org.` should also be NXDOMAIN, since `c.d.example.org.` is a non-wild empty non-terminal. Can you add this test case? It's a different case than the `something.r.c.d.example.org.` test case, since `r.c.d.example.org.` is not empty.
> > 
> > 
> > Hi @chrisohaver @SivaKesava1 Should`something.c.d.example.org.` also return NXDOMAIN? It looks to be the same case as `else.something.d.example.org`, it doesn't have a encloser under the wildcard `*.d.example.org` and is a multi-segment wildcard completion.
> 
> Yes, it should be an NXDOMAIN. Consider the domain names of the records as a tree, then follow the tree label by label from the root and stop when you can go further due to a non-match. If the node at which you stopped has a wildcard child, then you will perform wildcard synthesis; otherwise, you would return an NXDOMAIN.
> 
> ```
>     org
>       |
>    example
>       |
>       d
>      / \
>     c   *
>     |
>     r
> ```
> 
> When doing the matching, `something.c.d.example.org.` stops at the `c` node, and it does not have a wildcard child, so an NXDOMAIN. For `else.something.d.example.org,` it stops the matching at the `d` node, and it has a wildcard child, so it is a wildcard synthesis.
> I am not aware of the internal data structures used by CoreDNS, so I can not elaborate further, but the above explanation is based on how RFCs view the zone file.

@SivaKesava1 Thanks for the clarification, I need more changes for handling this above case.

**Comment by xdu31** [comment]
@chrisohaver @miekg @SivaKesava1 
1. Re-implementation: If the domain's terminal is neither the matching wildcard, nor a domian directly
under the wildcard, return NXDOMAIN
2. New test case: `something.c.d.example.org.`

File: plugin/file/lookup.go
Comment (chrisohaver) [inline plugin/file/lookup.go]: That this logic has essentially changed from `tr.Next(qname)` to `tr.Next(parts)`, seems like it might break empty non-terminal resolution.

Comment (chrisohaver) [reply plugin/file/lookup.go]: > That this logic has essentially changed from `tr.Next(qname)` to `tr.Next(parts)`, seems like it might break empty non-terminal resolution.

I don't see an existing test that validates queries for empty non-terminals should get a 'NOERROR' response. Should add it if one doesn't currently exist. e.g. in `lookup_test.go`...

`dbMiekNL` +=
```
test.ent	IN	A	139.162.196.79
```

and `dnsTestCases` +=
```
	{
		Qname: "ent.miek.nl.", Qtype: dns.TypeA,
		Ns: []dns.RR{
			test.SOA("miek.nl.	1800	IN	SOA	linode.atoom.net. miek.miek.nl. 1282630057 14400 3600 604800 14400"),
		},
	},
```

Comment (xdu31) [reply plugin/file/lookup.go]: Hi @chrisohaver I reverted back the tr.Next(qname) and added a new test case in lookup_test.go.

Comment (chrisohaver) [reply plugin/file/lookup.go]: Cool - thanks.  Looks like the other two wildcard related tests in azure and clouddns are passing now after that change too.

File: plugin/file/lookup.go
Comment (chrisohaver) [inline plugin/file/lookup.go]:
```suggestion
		// if the domain's terminal is neither the matching wildcard, nor a domain directly under the wildcard:
```

File: plugin/file/lookup.go
Comment (xdu31) [reply plugin/file/lookup.go]: Hi @chrisohaver, the idea is to iteratively find the next record that is the subdomain of the parts, it will finally find the last(longest) matching subdomain record.

For the zone file in `cornerCasesWildcard`
<pre>
     org
      |
   example
      |
      d
     / \
    c   *
    |
    r
</pre>

`example.org.`, ne is `*.d.example.org.`
`d.example.org.` ne is `*.d.example.org.`

`c.d.example.org.` ne is `r.c.d.example.org.`
`r.c.d.example.org.` ne is `r.c.d.example.org.`
...

`something.d.example.org.` ne is `*.d.example.org.` (inherent from `d.example.org.` as `*.d.example.org.` is not a parent of `something.d.example.org.`)
...

If number of labels of ne is greater than number of labels of the matching wildcard, we want to return NXDOMAINS

File: plugin/file/lookup.go
Comment (chrisohaver) [inline plugin/file/lookup.go:95]: Tests fail if the following record is added to the zone: `x.&.d.example.org.        IN      A       127.0.1.53`.

I think it's because of an assumption that a `*` element will always be ordered in the tree before any other element in the same sub domain (and `&` is sorted before `*`).  
E.g. the following record does not cause tests to fail: `x.+.d.example.org.        IN      A       127.0.1.53` presumably because `+` sorts after `*`.

File: plugin/file/wildcard_test.go
Comment (chrisohaver) [inline plugin/file/wildcard_test.go]: These IPs should be different from the wild card entry IP - So that an answer from these records can be discerned from an (incorrect) answer synthesized from the wildcard record.

**Review by chrisohaver — CHANGES_REQUESTED** [review]
Looks good.  The latest version of the logic feels more intuitive IMO. Thanks!

One minor comment about the test data, but other than that LGTM.

File: plugin/file/wildcard_test.go
Comment (xdu31) [reply plugin/file/wildcard_test.go]: @chrisohaver Change made.

**Review by chrisohaver — APPROVED** [review]
LGTM!  IMO an owner of the file plugin should review before merging.

**Comment by xdu31** [comment]
Hi @miekg @stp-ip @yongtang could you help review the PR?

**Review by yongtang — APPROVED** [review]
LGTM

File: plugin/file/lookup.go
Comment (miekg) [reply plugin/file/lookup.go:95]: sorting and tr.Next don't do what you think. 

and in general I think this PR just happens to implement what the tests hand it

**Comment by miekg** [comment]
meh, can't reopen old PRs.

anyways, finally have some time to comment/take a look. So the issue is/was that the current lookup algo should not have to change in response to this change. Adding a new element and taking Next() is thus probably not the right choice.

Also, I didn't think the tree is sorted and the layout of the elements depends on the insertion of them. I know, 'cause I tried doing this with NSEC and it did not work, although my tests all proved correct, but then I tried with larger zones and the whole thing fell apart.

**Comment by chrisohaver** [comment]
FYI, @xdu31 copying @miekg's comments from https://github.com/coredns/coredns/pull/4599#issuecomment-850348435

> anyways, finally have some time to comment/take a look. So the issue is/was that the current lookup algo should not have to change in response to this change. Adding a new element and taking Next() is thus probably not the right choice.
>
> Also, I didn't think the tree is sorted and the layout of the elements depends on the insertion of them. I know, 'cause I tried doing this with NSEC and it did not work, although my tests all proved correct, but then I tried with larger zones and the whole thing fell apart.

**Pull request merged**
